driver/main.c

```diff
@@ -1159,10 +1218,48 @@ static int record_event_consumer(struct ppm_consumer_t *consumer,
 #endif
 		args.buffer_size = min(freespace, delta_from_end) - sizeof(struct ppm_evt_hdr); /* freespace is guaranteed to be bigger than sizeof(struct ppm_evt_hdr) */
 		args.event_type = event_type;
-		args.regs = regs;
-		args.sched_prev = sched_prev;
-		args.sched_next = sched_next;
-		args.syscall_id = id;
+
+		if (event_datap->category == PPMC_SYSCALL) {
+			args.regs = event_datap->event_info.syscall_data.regs;
+			args.syscall_id = event_datap->event_info.syscall_data.id;
+		} else {
+			args.regs = NULL;
+			args.syscall_id = -1;
+		}
+
+		if (event_datap->category == PPMC_CONTEXT_SWITCH) {
+			args.sched_prev = event_datap->event_info.context_data.sched_prev;
+			args.sched_next = event_datap->event_info.context_data.sched_next;
+		} else {
+			args.sched_prev = NULL;
+			args.sched_next = NULL;
+		}
+
+		if (event_datap->category == PPMC_SIGNAL) {
+			args.signo = event_datap->event_info.signal_data.sig;
+
+			if (args.signo == SIGKILL) {
```

Comment:
Usually we put event parsing on filler function. Do you think is feasible here too? Like having raw event data on `args` and on filler parsing and add to event buffer.